Improve how data from Solr is queried when a shelflist.exporters.ItemsToSolr process is run

jthomale opened the pull request:
Fixes Issue #17 and otherwise improves how Solr is queried by `shelflist.search_indexes.ShelflistItemIndex` when an `ItemsToSolr` export process runs.

First, the problem with tons of HTTP connections being opened by querying Solr over and over is solved by allowing reuse of the same `pysolr.Solr` object across multiple `utils.solr.Queryset` instances.

Second, I've refactored `shelflist.search_indexes.ShelflistItemIndex` to make a minor improvement in how data is prepared for indexing. I replaced the four field `prepare_*` methods, where the same Solr query (to pull back data for the item represented by the object) was being sent once per field `prepare` method, with a single general object `prepare` method. So now Solr is queried once per item instead of four times per item.

jthomale commented:
@vphill @damonkelley @somexpert @jason-ellis @ldko -- Please take a look when you have a chance. Thanks!

@damonkelley I tried to make sure I didn't have extraneous whitespace this time, but I still may have missed some; maybe I _should_ start using vim instead of Sublime Text... :wink:

somexpert commented on django/sierra/shelflist/search_indexes.py:
```diff
@@ -1,65 +1,50 @@
+        self.prepared_data = super(ShelflistItemIndex, self).prepare(obj)
+        user_data_fields = ('shelf_status', 'inventory_notes', 'flags',
+                            'inventory_date')
+        for field in user_data_fields:
```
what is this `for` loop actually doing?

somexpert replied:
Unless I'm confused, I think the following `else` block needs to be indented another level

somexpert commented on django/sierra/shelflist/search_indexes.py:
```diff
@@ -1,65 +1,50 @@
+        for field in user_data_fields:
+            if getattr(obj, field, None) is not None:
+                break
+        else:
```
Think we can simplify the above 4 lines into this: `if all([hasattr(obj, field) for field in user_data_fields]):`

damonkelley replied:
It doesn't effect performance at all, but the brackets in the call to `all` are unnecessary.

This will also suffice
```python
if all(not hasattr(obj, field) for field in user_data_fields)
```

And a named function might lead to code that is a bit more intent revealing, while encapsulating the behavior

```python
def has_user_data_fields(self, obj):
     fields = ('shelf_status', 'inventory_notes', 'flags', 'inventory_date')
     return True if all(not hasattr(obj, f) for f in fields) else False
```

Which will give you this

```python
if self.has_user_data_fields(obj):
    try:
        # might blow up
```

vs.

```python
if all(not hasattr(obj, field) for field in user_data_fields):
    try:
        # might blow up
```

jthomale commented:
@somexpert Yes--much better! I spent more time than I'd like to admit yesterday trying to figure out a more concise way to do that. Totally blanked on the existence of both `hasattr` and `all`. Thank you!

damonkelley commented on django/sierra/shelflist/search_indexes.py:
```diff
@@ -1,65 +1,46 @@
+        index to add any existing values for these fields to the object
+        before it's re-indexed.
+        """
+        self.prepared_data = super(ShelflistItemIndex, self).prepare(obj)
```
Do you need to add an additional instance attribute? Or would `prepared_data` work just as well?

damonkelley commented on django/sierra/utils/solr.py:
```diff
@@ -46,21 +46,19 @@ def save(self, url=None, using='default', **kwargs):
+        if conn:
+            self._conn = conn
+        else:
+            self._conn = connect(url=url, using=using, **kwargs)
```
A trick I like to use when I need a default value in case one isn't passed in is this

```python
def __init__(self, url=None, using='default', page_by=100, conn=None, **kwargs):
    self._conn = conn or connect(url=url, using=using, **kwargs)
```
The `or` will short circuit if the first value is truthy.

Either way works the same though :smile:

damonkelley replied:
BTW,  not suggesting a change, just wanted to point it out.

jthomale commented:
@damonkelley @somexpert Addressed Damon's most recent comments (thanks, BTW) in the last few commits.

7483f69 makes the `self._conn` assignment in `utils.solr.Queryset.__init__` more concise. Generally speaking, I do like that style better because it's concise without obfuscating anything.

05a5779 breaks out the one-line boolean check for user data fields into a separate method. I also changed it from an `all()` check to make sure all user fields were not present to an `any()` check to see if any field was present: `any(hasattr(obj, field) for field in self.user_data_fields)`. I think that makes it plainer what I was trying to accomplish there. Ultimately I'm checking to make sure that the object doesn't have any user data on it before running the Solr query, and the resulting `if not self.has_any_user_data(obj)` in `prepare` reflects that better.  I also broke the `user_data_fields` out into a class attribute, since both `has_any_user_data` and `prepare` both need access to that list of fields, and that made more sense to me than passing the list from `prepare` to `has_any_user_data`.

2bb99d1 is a change that wasn't spurred by any comments, but I think falls in line with the goal of refactoring code to make it plainer what it's doing. The method I added to `utils.solr.Queryset` lets me try to get a single item from Solr without the try/except block in the `ShelflistItems.prepare` method, which was basically just checking to make sure the item from Solr existed. Now I just call the `Queryset.get_one` method. If it returns something then it merges that result with the data `obj` from Sierra--if not, it returns None, and it doesn't have to merge anything.

Finally, addressing this:

> Do you need to add an additional instance attribute? Or would `prepared_data` work just as well?

Good question. That's how the [Haystack docs](http://django-haystack.readthedocs.org/en/v2.4.1/searchindex_api.html#prepare-self-object) show overriding the `prepare` method for `haystack.SearchIndex`-derived classes (which `ShelflistItemIndex` is), so I assume that's probably what that class expects. It's been a while since I've looked at the Haystack code, though.

Thanks again for all the help!

damonkelley commented:
>Good question. That's how the Haystack docs show overriding the prepare method for haystack.SearchIndex-derived classes (which ShelflistItemIndex is), so I assume that's probably what that class expects. It's been a while since I've looked at the Haystack code, though.

Ah I see. It looked fishy to me because the method attaches/alters an instance variable **and** returns the value of that instance variable. But if that is what the docs say, then perhaps it is using the value somewhere else and relies on that context being available.

jthomale commented:
> Ah I see. It looked fishy to me because the method attaches/alters an instance variable and returns the value of that instance variable. But if that is what the docs say, then perhaps it is using the value somewhere else and relies on that context being available. ¯_(ツ)_/¯

Yeah--I agree. Both the example code they show and the `prepare` method (and related methods) on the Haystack `SearchIndex` class in their code do that--assign data to `self.prepared_data` and return it. There's a `SearchIndex.full_prepare` method that calls `self.prepare` and also assigns the results to `self.prepared_data`, which seems to make assigning the instance variable in `prepare` a little redundant, but ¯\\_(ツ)_/¯ ... I don't know if `prepare` gets called anywhere else or not.

damonkelley commented:
Good enough for me :+1:

somexpert commented:
Good call with using the `any` function. Lots of improvements to readability made. :thumbsup:
